### ISSUE TITLE
RHEL9 build issues (`setfiles: Could not set context for #file:  Invalid argument`)

### ISSUE THREAD

**Issue by yaroslav-gwit** [OSInside/kiwi#2519]
Good day. I hope you are all doing well.

## Problem description
I am trying to build a RHEL9 image, but it comes back with errors (debug logs included in the code block below).

## OS and Software information

* KIWI version: KIWI (next generation) version 10.0.4
* Operating system host version: AlmaLinux 8.9
* Operating system target version: RHEL 9
* Open Build Service version (N/A if not using OBS): N/A
* Koji version (N/A if not using Koji): N/A

Debug logs:
```log
[ INFO    ]: 11:47:43 | Calling config.sh script
[ DEBUG   ]: 11:47:43 | --------------out start-------------
[ DEBUG   ]: 11:47:43 | Configure image: [rhel-9]...
[ DEBUG   ]: 11:47:43 | --------------out stop--------------
[ INFO    ]: 11:47:43 | Processing SELinux file security contexts
[ DEBUG   ]: 11:47:43 | EXEC: [chroot /home/alma/OS_builds/rhel-9-hardened/build/image-root setfiles --help]
[ DEBUG   ]: 11:47:43 | EXEC: [chroot /home/alma/OS_builds/rhel-9-hardened/build/image-root setfiles -F -p -c /etc/selinux/targeted/policy/policy.33 -e /proc -e /sys -e /dev /etc/selinux/targeted/contexts/files/file_contexts /]
[ DEBUG   ]: 11:47:49 | EXEC: Failed with stderr: setfiles: Could not set context for /etc/NetworkManager/dispatcher.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/no-wait.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/pre-down.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/pre-up.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/cloud-init-azure-hook:  Invalid argument
setfiles: Could not set context for /run/motd.d:  Invalid argument
setfiles: Could not set context for /run/motd:  Invalid argument
setfiles: Could not set context for /var/lib/systemd/coredump:  Invalid argument
setfiles: Could not set context for /usr/lib/systemd/systemd-pstore:  Invalid argument
setfiles: Could not set context for /usr/lib/systemd/user/dbus-broker.service:  Invalid argument
setfiles: Could not set context for /usr/lib/systemd/systemd-network-generator:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/no-wait.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/pre-down.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/pre-up.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/11-dhclient:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/20-chrony-dhcp:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/20-chrony-onoffline:  Invalid argument
setfiles: Could not set context for /usr/bin/rpmdb:  Invalid argument
setfiles: Could not set context for /usr/libexec/nm-dispatcher:  Invalid argument
/ 100.0%: 


[ DEBUG   ]: 11:47:49 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/var/cache/kiwi]
[ DEBUG   ]: 11:47:49 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/var/cache/kiwi]
[ DEBUG   ]: 11:47:49 | EXEC: [umount -l /home/alma/OS_builds/rhel-9-hardened/build/image-root/var/cache/kiwi]
[ DEBUG   ]: 11:47:49 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/sys]
[ DEBUG   ]: 11:47:50 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/sys]
[ DEBUG   ]: 11:47:50 | EXEC: [umount -l /home/alma/OS_builds/rhel-9-hardened/build/image-root/sys]
[ DEBUG   ]: 11:47:50 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/dev]
[ DEBUG   ]: 11:47:50 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/dev]
[ DEBUG   ]: 11:47:50 | EXEC: [umount -l /home/alma/OS_builds/rhel-9-hardened/build/image-root/dev]
[ DEBUG   ]: 11:47:50 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/proc]
[ DEBUG   ]: 11:47:50 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root/proc]
[ DEBUG   ]: 11:47:50 | EXEC: [umount -l /home/alma/OS_builds/rhel-9-hardened/build/image-root/proc]
[ DEBUG   ]: 11:47:50 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root]
[ DEBUG   ]: 11:47:50 | EXEC: [mountpoint -q /home/alma/OS_builds/rhel-9-hardened/build/image-root]
[ DEBUG   ]: 11:47:50 | EXEC: [umount -l /home/alma/OS_builds/rhel-9-hardened/build/image-root]
[ DEBUG   ]: 11:47:50 | EXEC: [rmdir --ignore-fail-on-non-empty /home/alma/OS_builds/rhel-9-hardened/build/image-root/var/cache/kiwi]
[ DEBUG   ]: 11:47:50 | EXEC: [rmdir --ignore-fail-on-non-empty /home/alma/OS_builds/rhel-9-hardened/build/image-root/var/cache]
[ DEBUG   ]: 11:47:50 | EXEC: [rmdir --ignore-fail-on-non-empty /home/alma/OS_builds/rhel-9-hardened/build/image-root/var]
[ DEBUG   ]: 11:47:50 | EXEC: [rm -f /home/alma/OS_builds/rhel-9-hardened/build/image-root/etc/resolv.conf]
[ DEBUG   ]: 11:47:50 | EXEC: [rm -f /home/alma/OS_builds/rhel-9-hardened/build/image-root/etc/resolv.conf.kiwi /home/alma/OS_builds/rhel-9-hardened/build/image-root/etc/resolv.conf.sha /home/alma/OS_builds/rhel-9-hardened/build/image-root/etc/hosts.kiwi /home/alma/OS_builds/rhel-9-hardened/build/image-root/etc/hosts.sha]
[ ERROR   ]: 11:47:50 | KiwiCommandError: chroot: stderr: setfiles: Could not set context for /etc/NetworkManager/dispatcher.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/no-wait.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/pre-down.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/pre-up.d:  Invalid argument
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/cloud-init-azure-hook:  Invalid argument
setfiles: Could not set context for /run/motd.d:  Invalid argument
setfiles: Could not set context for /run/motd:  Invalid argument
setfiles: Could not set context for /var/lib/systemd/coredump:  Invalid argument
setfiles: Could not set context for /usr/lib/systemd/systemd-pstore:  Invalid argument
setfiles: Could not set context for /usr/lib/systemd/user/dbus-broker.service:  Invalid argument
setfiles: Could not set context for /usr/lib/systemd/systemd-network-generator:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/no-wait.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/pre-down.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/pre-up.d:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/11-dhclient:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/20-chrony-dhcp:  Invalid argument
setfiles: Could not set context for /usr/lib/NetworkManager/dispatcher.d/20-chrony-onoffline:  Invalid argument
setfiles: Could not set context for /usr/bin/rpmdb:  Invalid argument
setfiles: Could not set context for /usr/libexec/nm-dispatcher:  Invalid argument
/ 100.0%:
```

**Comment by yaroslav-gwit** [comment]
I've fixed the issue by switching `selinux` to permissive mode on my host system, aka:

```shell
sudo sed -i "s/SELINUX=enforcing/SELINUX=permissive/" /etc/selinux/config
sudo setenforce 0
```

But I don't know why this has showed up all of a sudden. Previous Alma8, Rocky8, Rhel8 image builds have finished just fine without the above config change.

**Comment by schaefi** [comment]
Hmm, do you think this is a problem in kiwi ? Unless you think the setfiles call done by kiwi:

```
EXEC: [chroot /home/alma/OS_builds/rhel-9-hardened/build/image-root setfiles -F -p -c /etc/selinux/targeted/policy/policy.33 -e /proc -e /sys -e /dev /etc/selinux/targeted/contexts/files/file_contexts /]
```

is wrong, I don't see how we could fix this.

```
setfiles: Could not set context for /etc/NetworkManager/dispatcher.d/no-wait.d:  Invalid argument
```

Isn't this more a problem in the security policy of selinux ?

**Comment by yaroslav-gwit** [comment]
My local SELinux was at fault there but it was rather unexpected, because it never happened before during my previous builds of `Alma8`, `Rocky8`, `Rhel8`, `Ubuntu`, `Debian`, etc.

So I am just wondering why it's different this time around (`RHEL8` and `RHEL9` are almost identical, apart from the repo locations, and a couple of package names) 🙂

Another question would be: is it okay to use ELv8 (I am using `Alma8` right now) in order to build the ELv9 images if the local SELinux is being used? I am asking, because it could (potentially) be incompatible across the two at some point.

**Comment by schaefi** [comment]
> My local SELinux was at fault there but it was rather unexpected,

Ah, so you are saying your local build host system is selinux enabled and that did not allow the selinux call to complete in the image build process ? I have seen several issues with selinux on the host preventing an image build to end successfully. The error message users get in this case are most often not leading to "it's selinux that forbids this..." thus really hard to find. One solution that does not require you to modify your host OS is to use kiwi's ```boxbuild``` rather than ```build```. See https://osinside.github.io/kiwi/plugins/self_contained.html for details. In this case your host only has to allow kvm instances to start up. The VM boxes within the actual build happens are designed to run image builds and prevents the issue you saw

> Another question would be: is it okay to use ELv8 (I am using Alma8 right now) in order to build the ELv9 images if the local SELinux is being used? I am asking, because it could (potentially) be incompatible across the two at some point.

I think @Conan-Kudo is best to answer this

**Comment by Conan-Kudo** [comment]
> Another question would be: is it okay to use ELv8 (I am using `Alma8` right now) in order to build the ELv9 images if the local SELinux is being used? I am asking, because it could (potentially) be incompatible across the two at some point.

@yaroslav-gwit This should be fine. The main thing though is that kiwi for EL8 is kiwi v9 in EPEL, so be mindful of some minor differences for the description definitions. If it works in kiwi v9 already, you should be good to go.

**Comment by Conan-Kudo** [comment]
But yes, you need to do `setenforce 0` before running a build, and then you can turn it back on afterward.

**Comment by yaroslav-gwit** [comment]
You might've missed the update I posted right under the original question, but it's okay. I've fixed the SELinux issue right after I realised that it was at fault (unfortunately I couldn't figure it out before posting this issue - the rubber ducky method kicked in automatically in my brain, I guess 😅).

Thanks for the link. I'll have read about the `boxbuild`, but I am already building my images under FreeBSD's `bhyve`, which doesn't support nested virtualisation unfortunately.

@Conan-Kudo thank you for the clarification. I am happy for you to close the issue.
You guys rock, as always 😉
